### PR TITLE
nixos/systemd-boot: Add support for an XBOOTLDR partition

### PR DESCRIPTION
**Pull request by sdht0**
## Description of changes

It is possible that the size of the ESP partition is very small (say 100MB), for e.g., when dual booting with Windows. Currently, systemd-boot install entries only to the ESP partition, which results in "Out of disk space" errors when trying to store multiple kernel generations. This PR adds support for an additional XBOOTLDR partition as described in the [Boot Loader Specification](https://uapi-group.org/specifications/specs/boot_loader_specification/#the-boot-loader-specification), which helps sidestep the issue.

The changes in systemd-boot nixos module were straighforward. However, to add tests, I had to make changes to the qemu and disk image modules. Perhaps there is a less invasive way by creating a custom VM only for these specific tests? Atleast the code in the disk image module will need to be replicated though. 

Thanks to @colemickens for encouraging me to further expand on his PR #226692.
Cc @ElvishJerricco @RaitoBezarius, can you take a look.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sdht0** [comment]
I separated out the changes to different modules and added comments.

I also added a release note for 23.11, although I'm not sure if this will be merged in time for the release.

 @ElvishJerricco @RaitoBezarius Can you take a look now? Thanks.

**Comment by sdht0** [comment]
@RaitoBezarius @ElvishJerricco Not sure if you've seen this yet. Maybe there is still time to get this merged for 23.11?

I think this is an important feature. I'm actually surprised it isn't already added. Perhaps most NixOS users don't dual boot with Windows or somehow work around the limited ESP partition size.

File: nixos/lib/make-disk-image.nix
Comment (ElvishJerricco) [inline nixos/lib/make-disk-image.nix:278]: This doesn't seem right. These are [Discoverable Partitions Specification](https://uapi-group.org/specifications/specs/discoverable_partitions_specification/) GUIDs, but you've used them for the partition GUIDs instead of the partition *type* GUIDs.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (ElvishJerricco) [inline nixos/modules/virtualisation/qemu-vm.nix:232]: In what circumstance is `config.boot ? loader` ever false?

File: nixos/tests/systemd-boot.nix
Comment (ElvishJerricco) [inline nixos/tests/systemd-boot.nix:18]: Use the module system to do merging, not `recursiveUpdate`. Just add `imports = [ common ];` instead.

**Review by ElvishJerricco — CHANGES_REQUESTED** [review]

File: nixos/lib/make-disk-image.nix
Comment (sdht0) [reply nixos/lib/make-disk-image.nix]: Right makes sense now. I reused ids from the next type.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (sdht0) [reply nixos/modules/virtualisation/qemu-vm.nix:265]: orfbog eval for nixos manual fails, for example.

File: nixos/tests/systemd-boot.nix
Comment (sdht0) [reply nixos/tests/systemd-boot.nix]: Updated.

**Review by ElvishJerricco — APPROVED** [review]

**Comment by sdht0** [comment]
Thanks @ElvishJerricco. Is there anything else I need to do to get this merged?

**Comment by sdht0** [comment]
Thanks for the ping! I'll take a look sometime later this week.

**Comment by ElvishJerricco** [comment]
@danielfullmer are you ok being added as a maintainer of these tests?

**Comment by RaitoBezarius** [comment]
> @danielfullmer are you ok being added as a maintainer of these tests?

I honestly was planning to remove Daniel from maintainers of systemd-boot (tests included) because I don't think he has been very active around them (which is fine).

@JulienMalka told me in private communications that he may be interested in taking over systemd-boot maintenance, at least, for the tests side.

**Comment by sdht0** [comment]
Ah right I copied the maintainer values too.

Actually, if you want, I can be a (co-)maintainer. I got a good enough understanding of how the systemd-boot installation and tests work and can help out with that atleast.

**Comment by sdht0** [comment]
Can we get this merged for 23.11? What are the remaining blockers?
@ElvishJerricco

**Comment by ElvishJerricco** [comment]
I'm ok with merging. @RaitoBezarius are we ok with the maintainers field in the test here?

**Comment by RaitoBezarius** [comment]
> I'm ok with merging. @RaitoBezarius are we ok with the maintainers field in the test here?

Nope, I don't think I have seen the people set in that maintainers field saying they were okay with that, cc @Enzime @danielfullmer.

Please remove them and add yourself.

**Comment by RaitoBezarius** [comment]
@JulienMalka as the new maintainer of the systemd boot builder script, you should review this.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (RaitoBezarius) [inline nixos/modules/virtualisation/qemu-vm.nix:1253]: Why is it RO here?

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (RaitoBezarius) [reply nixos/modules/virtualisation/qemu-vm.nix:265]: That's not normal, this can happen if you use the value of this inside of a manual-generated value I'd say.

**Review by RaitoBezarius — COMMENTED** [review]
I'm not a big fan of how intrusive supporting XBOOTLDR is, it seems fragile, and I am not sure any of the active maintainer will be able to support it, so please be certain that you will be around to actively tend to it if we end up merging this.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (RaitoBezarius) [reply nixos/modules/virtualisation/qemu-vm.nix:1253]: Arguably, this could be set by the bootloader directly, not sure if this is something the test framework has to do itself?

**Comment by JulienMalka** [comment]
I will review this change shortly

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix
Comment (JulienMalka) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:111]: I am thinking, maybe an assert to check that this mountPoint is actually define in `fileSystems` would help prevent mistakes here.

**Review by JulienMalka — CHANGES_REQUESTED** [review]
Thank you for your contribution! 
I think a nice to have for this kind of feature is having the `nixos-generate-config` pick it up automatically and generate the appropriate configuration, but that can come in a later change. Maybe a mention in the NixOS manual would be good too.

Are you intending on actively maintaining this feature? If yes, can you add yourself to the `maintainer-list.nix` file and to the `meta.maintainers` of your tests?

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix
Comment (ElvishJerricco) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:111]: I don't know if we care about this, but some people do allow `systemd-gpt-auto-generator` to create their `/boot` mount (and with this PR, they could do both their `/efi` and `/boot` mounts), meaning they wouldn't be in `fileSystems` at all.

Comment (sdht0) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:111]: I happen to use `systemd-gpt-auto-generator` as a matter of fact, so checking `fileSystems` won't work.

It's true that there is no reverting back if the systemd-boot config is messed up. I can setup an elaborate test in the python script to see if  `xbootldrMountPoint`/`efiSysMountPoint` are mountpoints and if the partition UUID is as expected, but that will need time.

Is such a test a blocker for the PR?

**Comment by sdht0** [comment]
> I'm not a big fan of how intrusive supporting XBOOTLDR is, it seems fragile

@RaitoBezarius Can you say more on why you think it's fragile. The changes are a little intrusive, yes. especially in `make-disk-image.nix`, which need to setup two boot partitions instead of one. However, I think the code is well isolated behind the configured flag. I also think anyone should be able to quickly catch on what's happening if they understand the rest of the code.

And yeah I'm planning to stick around and help out as needed.

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py
Comment (JulienMalka) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:348]: Hmm unless I am wrong, code in finally block will run unconditionally, so if `xbooldrMountPoint` is not available, it will not copy the entries, not install the bootloader, and then still clean the boot partition.

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix
Comment (JulienMalka) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:111]: I think it should be a blocker yes. As it is, an uncaring user might end up making their machine unbootable, we should have guards against that. Using an assertion doesn't seem to be the good implementation if there is a `systemd-gpt-auto-generator` usecase but something in the backend should be possible.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (sdht0) [reply nixos/modules/virtualisation/qemu-vm.nix:1253]: > Why is it RO here?

Not sure. I just mirrored the `/boot` entry above this.

> Arguably, this could be set by the bootloader directly, not sure if this is something the test framework has to do itself?

I think it's better to do it here in one place. If `setupXbootldr` is true, the VM should mount all the boot partitions created in the disk image.

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py
Comment (sdht0) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:348]: Ok that's a good catch that I didn't notice. At the very least, cleanup_esp() should then be inside install_bootloader() not here. I'll make that change.

Let's move the conversation from the other comment here. What do you have in mind when you say "something in the backend"? On my system, because of systemd auto mounting, the actual `/efi` and `/boot` are not even mounted until those partitions are accessed:
```bash
λ cat /proc/mounts | grep -E 'efi|boot'
efivarfs /sys/firmware/efi/efivars efivarfs rw,nosuid,nodev,noexec,relatime 0 0
systemd-1 /boot autofs rw,relatime,fd=157,pgrp=1,timeout=120,minproto=5,maxproto=5,direct,pipe_ino=4420 0 0
systemd-1 /efi autofs rw,relatime,fd=158,pgrp=1,timeout=120,minproto=5,maxproto=5,direct,pipe_ino=4422 0 0
```
So the partition I had in mind will first need to trigger a mount before it can check the partition id. Maybe you can think of a simpler test?

Comment (JulienMalka) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:348]: I think just catching the cases where the `xbootldrMountPoint` is not available, and displaying a helpful error message is good enough.

Comment (JulienMalka) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:348]: For me the two problems were related, solving it either by catching the user mistake at the module level or in the bootloader backend is fine :)

Comment (JulienMalka) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py]: > So the partition I had in mind will first need to trigger a mount before it can check the partition id. Maybe you can think of a simpler test?

Not sure that there is anything simplier than triggering the mount and checking for partition id.

Comment (sdht0) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py]: Alright so three changes:
* Moved the cleanup code. Now if entries are not written to `xbootldrMountPoint` for whatever reason, the cleanup code will not run.
* Added assertions to make sure the mount points are absolute paths.
* Added a simple check to ensure `xbootldrMountPoint`/`efiSysMountPoint` are mountpoints.

 @JulienMalka  With these I think we get most of the way there. Do you think these are sufficient to unblock the PR?

Comment (sdht0) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py]: I cannot seem to find a single robust way to get the partition GUID. It might even turn out to be fragile and break existing setups.

Note that systemd-boot itself checks for the correct partition guid on an install or update. The only remaining case where this might fail is for a user on an existing setup sets `xbootldrMountPoint` to a mounted partition path that is not the XBOOTLDR partition and there is no systemd-boot install or update to be done. I've updated the config description to remind the users about expected mount points..

Comment (JulienMalka) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py]: I think the mitigations you implemented are already pretty good, I need time to review them and test them properly though :)

Comment (sdht0) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py]: Sounds good. I tested locally by deliberately trying to trigger all the assertions and checks, both with my [nixos-config](https://github.com/NixOS/nixpkgs/pull/226692#pullrequestreview-1651729044) and in the nixos tests.

Comment (sdht0) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py]: @JulienMalka Did you get a chance to look at this?

Comment (JulienMalka) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py]: Yes, I think this looks good, thanks. Unfortunately, #263442 got merged in the meanwhile, meaning that substantial changes have been done to the `systemd-boot-builder.py` file and that you have to resolve the conflicts.

Comment (sdht0) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py]: @JulienMalka fixed conflicts and moved release notes to 24.05.

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py
Comment (JulienMalka) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:351]: What is the effect of this if placed here?

Comment (JulienMalka) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:351]: Sorry my eyes got tricked by the way github was showing the diff, I thought it was at the end of the `install_bootloader()` function. This is perfectly fine. I'll finish the review tomorrow when I am less tired.

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix
Comment (ElvishJerricco) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:45]: Should go to stderr.

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix
Comment (ElvishJerricco) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:278]: nitpick

```suggestion
        message = "The ESP mount point '${efi.efiSysMountPoint}' must be an absolute path";
      }
      {
        assertion = cfg.xbootldrMountPoint == null || (hasPrefix "/" cfg.xbootldrMountPoint);
        message = "The XBOOTLDR mount point '${cfg.xbootldrMountPoint}' must be an absolute path";
```

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (ElvishJerricco) [reply nixos/modules/virtualisation/qemu-vm.nix:265]: I'd rather get to the bottom of it than leave this incredibly strange, and likely wrong-ish condition.

**Review by JulienMalka — APPROVED** [review]
Changes to the `systemd-boot-builder.py` and `systemd-boot.nix` files seem good as far as I am concerned. 
Thanks you for submitting this change and taking the time to get it to completion, even though you had big conflicts to change with the migration to bootspec.
Still some minors changes to address from other reviewers and conflicts with the release notes and I think we can merge that.
As a minor piece of advice, please do not make changes to code that is not directly related to your PR, even if cosmetic or very minor, it makes it more difficult to review.

**Comment by JulienMalka** [comment]
@ofborg test systemd-boot

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (sdht0) [reply nixos/modules/virtualisation/qemu-vm.nix]: From what I can tell, the error is likely because the manual release code only loads a restricted set of modules that does not include the `boot.loader` module.

The solution I can think of is to introduce another config option in `qemu-vm.nix` instead of trying to directly use the config option from `systemd-boot.nix`.

Will need help with a better way if the above is not acceptable.

**Comment by sdht0** [comment]
@JulienMalka @ElvishJerricco I addressed the comments in a new commit to make it easy to review again. Once ready to merge, I'll do a final squash.

**Comment by RaitoBezarius** [comment]
@JulienMalka ping ^

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (sdht0) [reply nixos/modules/virtualisation/qemu-vm.nix]: @RaitoBezarius: I think the above is the only blocker. @JulienMalka already reviewed the other changes. If you or @ElvishJerricco can review the latest commit, this PR should be ready to merge.

**Comment by JulienMalka** [comment]
Friendly ping @ElvishJerricco, your currently standing review is the last blocker for this.

**Comment by nyabinary** [comment]
> Friendly ping @ElvishJerricco, your currently standing review is the last blocker for this.

It also needs a rebase :3

**Comment by ElvishJerricco** [comment]
@sdht0 I've gone ahead and rebased, resolved conflicts, and cleaned up the git history in my branch: https://github.com/NixOS/nixpkgs/compare/master...ElvishJerricco:nixpkgs:systemd-boot-xbootldr

The only significant conflict was a slight refactor of `systemd-boot-builder.py`, so I edited the changes to adhere to that refactor. Do you mind if I push that to this PR?

**Comment by sdht0** [comment]
@ElvishJerricco nit: maybe remove the stray /EFI capitalization if it's lowercase everywhere else, otherwise looks good. Alright. let's do this!

**Comment by ElvishJerricco** [comment]
> maybe remove the stray /EFI capitalization if it's lowercase everywhere else

Fair enough. Though, I *would like* to have that change eventually, because if you load a driver for a case-sensitive file system, it does actually matter. We can leave it for later though

**Review by ElvishJerricco — APPROVED** [review]
@JulienMalka It looks good to me now. Feel free to double check since I rebased it.

@RaitoBezarius 

> I'm not a big fan of how intrusive supporting XBOOTLDR is, it seems fragile, and I am not sure any of the active maintainer will be able to support it, so please be certain that you will be around to actively tend to it if we end up merging this.

Are your concerns here sufficiently alleviated?

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix
Comment (sdht0) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:64]: There are still uppercase paths here. At the risk of delaying this PR more, I introduced a variable for this path: https://github.com/NixOS/nixpkgs/commit/e28fb92c82b36875c02d000f68c9ac642da4c0d1. Feel free to cherry-pick this commit if you want. Or we can just fix these paths as of now.

There's also one last use at line 89 of the python file, but we ignore the case there anyways.

**Comment by colemickens** [comment]
@ofborg eval

**Comment by colemickens** [comment]
Meant to send this comment last night, but I tested this (the latest variable commit) on a new Rock5b install with split efi and (ext)boot and it worked perfectly.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (RaitoBezarius) [inline nixos/modules/virtualisation/qemu-vm.nix:287]:
```suggestion
    partitionTableType = selectPartitionTableLayout { inherit (cfg) useDefaultFilesystems useEFIBoot useXbootldr; };
```

**Comment by RaitoBezarius** [comment]
> @JulienMalka It looks good to me now. Feel free to double check since I rebased it.
> 
> @RaitoBezarius
> 
> > I'm not a big fan of how intrusive supporting XBOOTLDR is, it seems fragile, and I am not sure any of the active maintainer will be able to support it, so please be certain that you will be around to actively tend to it if we end up merging this.
> 
> Are your concerns here sufficiently alleviated?

No, the QEMU VM is still too much from my personal perspective, but I won't block on that, if we have any issue with this, I will just ping the maintainer at that time and if we cannot get it to work, I will schedule it for removal.

@Nikstur and I have been planning a bunch of overhauls for the QEMU VM framework interface, and we would like to keep the module modifications to a minimal until we can get started.

In my ideal world, XBOOTLDR should just be normal operations of ESP + a weird /boot partition and all of that has nothing to do in internals like qemu-vm.nix, I made my point, I unfortunately don't have much time to stick around.

**Comment by nikstur** [comment]
I'm really quite opposed to the changes to the qemu-vm.nix module and I also believe they are not needed.

> I also think anyone should be able to quickly catch on what's happening if they understand the rest of the code.

I do not feel that way. I've read and changed the qemu-vm.nix code many times and its convoluted and confusingly mixes concepts all the time. It is extremely hard to change and even harder to change without breaking things. I feel like your change would pile on to this and make the issue worse.

Additionally, XBOOTLDR is generally not very useful for VMs because you typically can just make the ESP as large as you need in a virtual image. The only real value then lies in testing. However, I don't want to treat qemu-vm as the kitchen sink where we implement all sorts of testing features.

Instead of changing qemu-vm.nix, you should do:

- Create the image with an XBOOTLDR partition in the test that requires it by directly calling make-disk-image 
- Tell the test driver to use this image instead by pointing NIX_DISK_IMAGE to your image. You can find an example of how this works here: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/appliance-repart-image.nix#L100

You could put this in a fixture so that it can be reused by others if that's a concern.

**Comment by sdht0** [comment]
Deleted the branch by an incorrect push, sorry.

@nikstur @RaitoBezarius After many struggles, I've managed to figure out how to avoid changed in qemu-vm. Can you please take a look at https://github.com/NixOS/nixpkgs/pull/285401/commits/7afa4bcdf9ed42fb6538055db5901910a789e5e4#diff-f71ea5f910763bce18731f20cfb7fa3c6f31d4ebecf3a37c75cf410858b5bb53.

@JulienMalka @ElvishJerricco the rest of the changes remain the same in the new PR, but please do take another look.

**Pull request closed without merging**
